gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/AccountController.groovy

```diff
@@ -0,0 +1,97 @@
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.security.access.prepost.PostFilter
+import org.springframework.security.access.prepost.PreAuthorize
+import org.springframework.web.bind.annotation.DeleteMapping
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.PostMapping
+import org.springframework.web.bind.annotation.PutMapping
+import org.springframework.web.bind.annotation.RequestBody
+import org.springframework.web.bind.annotation.RequestParam
+import org.springframework.web.bind.annotation.RestController
+
+@RestController
+@Api
+class AccountController {
+
+  @Autowired
+  ClouddriverService clouddriverService
+
+  @GetMapping('/account/{accountName}')
+  @ApiOperation('Looks up an account definition by name.')
+  @PreAuthorize('''hasPermission(#accountName, 'ACCOUNT', 'READ')''')
+  Map getAccountByName(
+    @ApiParam('Name of account to get details.')
+    @PathVariable String accountName
+  ) {
+    clouddriverService.getAccountDefinition(accountName)
+  }
+
+  @GetMapping('/accounts/{accountType}')
+  @ApiOperation('Looks up account definitions by type.')
+  @PostFilter('''hasPermission(filterObject.name, 'ACCOUNT', 'READ')''')
+  List<Map> getAccountsByType(
+    @ApiParam(value = 'Value of the "@type" key for accounts to search for.', example = 'kubernetes')
+    @PathVariable String accountType,
+    @ApiParam('Maximum number of entries to return in results. Used for pagination.')
+    @RequestParam OptionalInt limit,
+    @ApiParam('Account name to start account definition listing from. Used for pagination.')
+    @RequestParam Optional<String> startingAccountName
+  ) {
+    clouddriverService.getAccountDefinitionsByType(accountType, limit.isPresent() ? limit.getAsInt() : null, startingAccountName.orElse(null))
+  }
+
+  @PostMapping('/account')
+  @ApiOperation('Creates a new account definition.')
+  @PreAuthorize('@authController.isAdmin()')
+  Map createAccount(
+    @ApiParam('Account definition body including a discriminator field named "@type" with the account type.')
+    @RequestBody Map accountDefinition
+  ) {
+    clouddriverService.createAccountDefinition(accountDefinition)
+  }
+
+  @PutMapping('/account')
+  @ApiOperation('Updates an existing account definition.')
+  @PreAuthorize('@authController.isAdmin()')
+  Map updateAccount(
+    @ApiParam('Account definition body including a discriminator field named "@type" with the account type.')
+    @RequestBody Map accountDefinition
+  ) {
+    clouddriverService.updateAccountDefinition(accountDefinition)
+  }
+
+  @DeleteMapping('/account/{accountName}')
+  @ApiOperation(value = 'Deletes an account definition by name.',
+    notes = 'Deleted accounts can be restored via the update API. Previously deleted accounts cannot be "created" again to avoid conflicts with existing pipelines.')
```

Comment:
I'm curious about this - in a setup where an account isn't a cluster but instead namespace scoped (think multi-tenant), at scale this could become quite unwieldy to manage as namespaces come and go whereas clusters more often than not don't.

I wonder if it would make sense to have a force delete option?

Reply:
I think this makes sense. I initially used this behavior due to keeping an update history of account info, but given the lack of foreign key between the history table and the accounts table, I think this might make sense to update to do a proper delete.